### PR TITLE
4172 aggregate with  buffer copy

### LINKED ISSUE #4172: run `DiceMetric.aggregate()` more than once produce inaccurate results
**Describe the bug**
when you apply `aggregate()` a `DiceMetric` object more than once, it will treat `nan` as `0` in the calculation, resulting in a different and inaccurate mean

**To Reproduce**
[https://gist.github.com/yiyixuxu/861d731d682fbb30fa281990f5e1980f](url)

```python

from monai.metrics import DiceMetric
from monai.losses.dice import *  # NOQA
import torch
input1 = torch.tensor(
    [[[[0, 0],
       [0, 0]], 
      [[1, 1],
       [0, 1]],
     [[0, 0],
      [1, 0]]]])
target1 = torch.tensor(
    [[[[1., 0.],
       [1., 1.]],
      
      [[0., 1.],
       [0., 0.]],
     
     [[0., 0.],
      [0., 0.]]]])
dice_metric = DiceMetric(include_background=True, reduction="mean")
dice1 = dice_metric(y_pred=input1, y=target1)
print(f'dice1: {dice1}')
>>> dice1: tensor([[0.0000, 0.5000,    nan]])
print(f'dice_metric.aggregate().item():{dice_metric.aggregate().item()}')
>>>dice_metric.aggregate().item():0.25
print(f'dice_metric.aggregate().item():{dice_metric.aggregate().item()}')
>>>dice_metric.aggregate().item():0.1666666716337204
``
**Expected behavior**
It should have consistent results as `0.25` 


**Additional context**
this happens because the `do_metric_reduction` method modify the cache of buffer data in place;  
https://github.com/Project-MONAI/MONAI/blob/dev/monai/metrics/utils.py#L45

### LINKED ISSUE #4176: `reduction` argument for `DiceMetric.aggregate()`?
**Is your feature request related to a problem? Please describe.**

if I want to track dice score on more than one dimensions,  for example, if I need both average dice per class and average dice. it seems like my only options are 
(1) create multiple DiceMetric object
(2) create `1` DiceMetric object to get average dice per class, then aggregate manually on top of that 

if there is a better way, please let me know! 
otherwise is it possible to let user pass a `reduction` argument to `aggregate()` method instead of `__init__`? 

**Describe alternatives you've considered**
```python
def aggregate(self, reduction=None):
  data = self.get_buffer()
  if not isinstance(data, torch.Tensor):
    raise ValueError("the data to aggregate must be PyTorch Tensor.")

  # do metric reduction
  if reduction is None: 
    reduction = self.reduction 
  f, not_nans = do_metric_reduction(data, self.reduction)
  return (f, not_nans) if self.get_not_nans else f
```

### PR DESCRIPTION
**Pull request by wyli**
Signed-off-by: Wenqi Li <wenqil@nvidia.com>

Fixes #4172
Fixes #4176

### Description
copy the buffer when aggregating.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Review by Nic-Ma — APPROVED** [review]
Thanks for the quick fix.

**Comment by Nic-Ma** [comment]
Mark: may also cover https://github.com/Project-MONAI/MONAI/issues/4176 in this PR.

Thanks.
